### ISSUE TITLE
Should TunnelCommunity be a subclass of TrustchainCommunity?

### ISSUE THREAD

**Issue by devos50** [Tribler/tribler#3335]
I'm working on the code of the `TunnelCommunity` and I wonder whether it makes more sense if we make this community a subclass of `TrustChainCommunity`, essentially building it on top of the TrustChain. They already share the same candidates and the design right now is quite fragile. When interacting with this community, we have to rely on the Tribler notification system.

Thoughts?

**Comment by synctext** [comment]
No objections from my side, just questions.
We're going towards a single walker and move beyond the Dispersy isolated community model.
What is the impact on the upcoming IPv8 replacement of Dispersy/Tunnel/Tribler stack. How future-proof is this effort? (no problem if it's a two-day job obviously). If we want to go wild and implement "Secure Homomorphic Partitioning" from Quinten his thesis (#2571 related, fixed block sizes, fixed payout units).

**Comment by devos50** [comment]
@synctext it doesn't really influence existing or future functionality. It's just another way of organizing the relations between communities in the code. For now, we've been developing Trustchain as a separate community. My suggestion is that we reorganize the subclass tree so we build the `TunnelCommunity` on top of Trustchain. Note that after this change, we still have a separation between these two communities. 

This would take at most a few hours of development time I think and should make it a bit easier to work with the `TunnelCommunity` in combination with Trustchain. It also helps when creating Gumby experiments.

**Comment by qstokkink** [comment]
Firstly,  code-wise: neither the complexity of the code nor the length will change significantly and I see no complications with IPv8 or possible other future gimmicks.
Secondly, it will become slightly harder to debug and isolate the communities.
Thirdly, you can no longer opt-out of the (privacy-weakening) TrustChain when using the TunnelCommunity.

If I were a Tor engineer, the third point would be unacceptable. However, as we want to start providing the TunnelCommunity as a "paid service"/currency, I think that this lock-in is acceptable.

**tl;dr** I will hesitantly agree with this proposition.

**Comment by devos50** [comment]
It turns out that the tunnel community and TrustChain community have different master member identities and thus, cannot easily be merged without breaking backwards compatibility since they are two truly separate communities. I'd rather not change the identifiers again to avoid updating infrastructure since we already have various versions of the TrustChain community running.

We should have a discussion about this soon.
